pkg/controllers/provisioningv2/rke2/provisioningcluster/template.go

```diff
@@ -273,16 +301,59 @@ func machineDeployments(cluster *rancherv1.Cluster, capiCluster *capi.Cluster, d
 		machinePoolNames[machinePool.Name] = true
 
 		var (
-			machinePoolName = name.SafeConcatName(cluster.Name, machinePool.Name)
-			infraRef        corev1.ObjectReference
+			machineDeploymentName     = name.SafeConcatName(cluster.Name, machinePool.Name)
+			existingMachineDeployment = false
+			infraRef                  corev1.ObjectReference
 		)
 
+		ls, err := labels.Parse(fmt.Sprintf("%s=%s,%s=%s", capi.ClusterLabelName, cluster.Name, rke2.RKEMachinePoolNameLabel, machinePool.Name))
+		if err != nil {
+			return nil, err
+		}
+
+		md, err := capiMachineDeployments.List(cluster.Namespace, ls)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(md) > 1 {
+			return nil, fmt.Errorf("multiple machinedeployments found with labels %s", fmt.Sprintf("%s=%s,%s=%s", capi.ClusterLabelName, cluster.Name, rke2.RKEMachinePoolNameLabel, machinePool.Name))
```

Comment:
I see one potential issue here - every machine deployment for every machine pool in every cluster lives in fleet-default. I do not believe that we stop an end user from using the same name for multiple machinepools in the same cluster.

Reply:
https://github.com/rancher/rancher/pull/36614/files#diff-4ebed3bcbfb9e2d6818b0f67b6dfb8bf6b480340940582779c379f585af0b431R298-R300

We do check for duplicate machinePool names within the same cluster.

Reply:
This is a bug on the UI side - the backend doesn't allow dupe machine pool names but the UI let's you save a cluster config that contains dupe machine pool names and then the cluster immediately goes into an error state. resolving this